### PR TITLE
Lower symbolic bool not to torch.sym_not

### LINKED ISSUE #143157: is it desirable to assume complex expressions in torch._check to be true
### 🚀 The feature, motivation and pitch

For the following code,
```python
import torch
torch._dynamo.config.capture_scalar_outputs = True
def f(x):
    c = x.item()
    return torch.ones(c*2)

out = torch.compile(f, fullgraph=True)(torch.tensor(3, dtype=torch.int64))
```

I'm getting an data dependent error: 
```
GuardOnDataDependentSymNode: Could not guard on data-dependent expression Eq(2*u0, 0) (unhinted: Eq(2*u0, 0)).  (Size-like symbols: none)
```

So ideally I could just add a torch._check to unblock myself:
```python
import torch
torch._dynamo.config.capture_scalar_outputs = True
def f(x):
    c = x.item()
    torch._check(not (c*2 == 0))
    return torch.ones(c*2)

out = torch.compile(f, fullgraph=True)(torch.tensor(3, dtype=torch.int64))
```
But the error is still there. After talking to @pianpwk, it seems that it's difficult to reason the range of`not equal` operator and more complicated  ones like %, etc.

So I'm wondering if it's desirable that we can assume the information we get from users e.g. via torch._check is correct. This would be mostly useful in draft mode export i think, where correctness doesn't matter that much.

### Alternatives

_No response_

### Additional context

_No response_

cc @chauhang @penguinwu @avikchaudhuri @gmagogsfm @zhxchen17 @tugsbayasgalan @angelayi @suo @ezyang @pianpwk

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.16.0) (oldest at bottom):
* __->__ #186043

Dynamo handles Python `not` as `operator.not_` during bytecode tracing. For
`SymNodeVariable` values that represent symbolic booleans, the old lowering also
emitted `operator.not_` into the FX graph. Fake evaluation of that node forced
the SymBool through its example hint, so `torch._check(not expr)` could be
constant-folded away instead of installing the symbolic assertion. Later uses of
`expr` then failed with a data-dependent guard.

Lower symbolic boolean negation to `torch.sym_not` instead. That keeps the
negated symbolic expression in the graph and lets `torch._check` preserve the
assertion. Non-boolean symbolic values keep the existing `operator.not_` path,
which implements normal Python truthiness.

Non-strict export executes user Python directly rather than through Dynamo
bytecode tracing, so Python `not` still reaches `SymBool.__bool__` before
`torch._check` can see it. Add an error hint for that path pointing users to
`torch.sym_not()` or an equivalent symbolic comparison.

Fixes #143157
Generated by my agent

Test Plan:
- python test/dynamo/test_misc.py MiscTests.test_torch_check_symbool_python_not -v
- python test/export/test_export.py TestExport.test_torch_check_symbool_python_not_strict TestExport.test_torch_check_symbool_python_not_nonstrict_error -v
- lintrunner -a

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @jataylo @azahed98

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says the Dr. CI failures were caused by translation validation not knowing how to translate the new torch.sym_not ShapeEnv node into Z3, so the assertion argument collapsed to a Python bool. I added the torch.sym_not -> z3.Not mapping in torch/fx/experimental/validator.py and verified the failing dynamic-shapes test plus the original Dynamo/export regression tests locally. lintrunner is blocked by an unrelated Pyrefly error in torch/nn/modules/loss.py.

**Comment by jansel** [comment]
My agent says the remaining Dr. CI dynamic-shapes failures were unexpected successes: the baddbmm scalar-arg dynamic-shapes test now passes after the torch.sym_not/Z3 validator fix, but it still had a stale expectedFailureDynamic marker. I removed that stale xfail and verified the baddbmm dynamic-shapes test, the scalar-arg parametrized tests, the original Dynamo/export regression tests, and lintrunner -a locally.

**Review by anijain2305 — APPROVED** [review]

**Review by mlazos — APPROVED** [review]

**Comment by jansel** [comment]
My agent says the post-rebase CI failures all came from test/dynamo/test_misc.py: the PR had removed expectedFailureDynamic, but current main added two decorators that still use it, causing import-time NameError across 62 checks. I kept the now-required import and updated the symbolic-not graph assertion for current main's aten._assert_scalar lowering; the focused Dynamo, export, and dynamic-shapes tests plus lintrunner -a pass locally.
